File: src/replication.zig
const std = @import("std");
const helpers = @import("helpers.zig");
const state_mod = @import("state.zig");
const posix = helpers.posix;
const c = helpers.c;
const log = helpers.log;
const FsState = state_mod.FsState;

pub fn replWorkerLoop(state: *FsState) void {
    while (!state.shutdown.load(.acquire)) {
        const work = state.repl_log.dequeueNext() orelse break;

        var backoff_ns: u64 = 1_000_000_000;
        const max_backoff_ns: u64 = 300_000_000_000;

        while (!state.shutdown.load(.acquire)) {
            const result = switch (work.op) {
                .put => replicatePut(state, work.path),
                .delete => replicateDelete(state, work.path),
            };

            if (result) |_| {
                state.repl_log.markCompleted(work.id);
                break;
            } else |err| {
                log.err("replication error for {s}: {}, retrying in {d}s", .{
                    work.path,
                    err,
                    backoff_ns / 1_000_000_000,
                });
                std.Thread.sleep(backoff_ns);
                backoff_ns = @min(backoff_ns * 2, max_backoff_ns);
            }
        }
    }
}

pub fn replicatePut(state: *FsState, rel_path: []const u8) !void {
    const backing_path = try std.fs.path.join(state.allocator, &.{ state.backing_dir, rel_path });
    defer state.allocator.free(backing_path);
    const replica_path = try std.fs.path.join(state.allocator, &.{ state.replica_dir, "files", rel_path });
    defer state.allocator.free(replica_path);

    const backing_stat = posix.fstatat(posix.AT.FDCWD, backing_path, posix.AT.SYMLINK_NOFOLLOW) catch |err| switch (err) {
        error.FileNotFound => return,
        else => return err,
    };

    if (backing_stat.mode & posix.S.IFMT == posix.S.IFLNK) {
        try ensureParentDir(replica_path);
        std.fs.deleteFileAbsolute(replica_path) catch |err| switch (err) {
            error.FileNotFound => {},
            else => return err,
        };
        const backing_z = try state.allocator.dupeZ(u8, backing_path);
        defer state.allocator.free(backing_z);
        const replica_z = try state.allocator.dupeZ(u8, replica_path);
        defer state.allocator.free(replica_z);
        var link_buf: [std.fs.max_path_bytes]u8 = undefined;
        const target = posix.readlinkat(posix.AT.FDCWD, backing_z, &link_buf) catch return error.ReadLinkFailed;
        posix.symlinkat(target, posix.AT.FDCWD, replica_z) catch return error.SymlinkFailed;
        return;
    }

    const sum_backing = try std.fmt.allocPrint(state.allocator, "{s}.sum", .{backing_path});
    defer state.allocator.free(sum_backing);
    const sum_replica = try std.fmt.allocPrint(state.allocator, "{s}.sum", .{replica_path});
    defer state.allocator.free(sum_replica);

    if (helpers.readSumFile(state.allocator, sum_backing)) |stored_hex| {
        defer state.allocator.free(stored_hex);
        if (helpers.computeBlake3(backing_path)) |computed_hex| {
            if (!std.mem.eql(u8, &computed_hex, stored_hex)) {
                log.warn("replication: skipping {s} — backing file does not match .sum (possible corruption)", .{rel_path});
                return;
            }
        } else |_| {}
    } else |_| {}

    try ensureParentDir(replica_path);
    try copyFileWithSync(backing_path, replica_path);

    copyFileWithSync(sum_backing, sum_replica) catch |err| switch (err) {
        error.FileNotFound => {},
        else => return err,
    };

    const stat_info = posix.fstatat(posix.AT.FDCWD, backing_path, 0) catch return;
    const mode: posix.mode_t = stat_info.mode & 0o7777;
    posix.fchmodat(posix.AT.FDCWD, replica_path, mode, 0) catch {};

    const replica_z = state.allocator.dupeZ(u8, replica_path) catch return;
    defer state.allocator.free(replica_z);
    _ = c.chown(replica_z.ptr, stat_info.uid, stat_info.gid);
}

pub fn replicateDelete(state: *FsState, rel_path: []const u8) !void {
    const replica_path = try std.fs.path.join(state.allocator, &.{ state.replica_dir, "files", rel_path });
    defer state.allocator.free(replica_path);
    const sum_replica = try std.fmt.allocPrint(state.allocator, "{s}.sum", .{replica_path});
    defer state.allocator.free(sum_replica);

    std.fs.deleteFileAbsolute(replica_path) catch |err| switch (err) {
        error.FileNotFound => {},
        else => return err,
    };
    std.fs.deleteFileAbsolute(sum_replica) catch |err| switch (err) {
        error.FileNotFound => {},
        else => return err,
    };

    if (!state.no_remote_mkdir) {
        const replica_files_root = try std.fs.path.join(state.allocator, &.{ state.replica_dir, "files" });
        defer state.allocator.free(replica_files_root);
        removeEmptyParentDirs(replica_path, replica_files_root);
    }
}

pub fn copyFileWithSync(src_path: []const u8, dst_path: []const u8) !void {
    const src = try std.fs.openFileAbsolute(src_path, .{});
    defer src.close();

    var tmp_path_buf: [std.fs.max_path_bytes]u8 = undefined;
    const tmp_path = std.fmt.bufPrint(&tmp_path_buf, "{s}.tmp", .{dst_path}) catch return error.NameTooLong;

    const dst = try std.fs.createFileAbsolute(tmp_path, .{});

    var ok = false;
    defer {
        if (!ok) {
            dst.close();
            std.fs.deleteFileAbsolute(tmp_path) catch {};
        }
    }

    var buf: [64 * 1024]u8 = undefined; // 64 KB
    while (true) {
        const n = try src.read(&buf);
        if (n == 0) break;
        try dst.writeAll(buf[0..n]);
    }
    try dst.sync();
    dst.close();
    ok = true;

    std.fs.renameAbsolute(tmp_path, dst_path) catch |err| {
        std.fs.deleteFileAbsolute(tmp_path) catch {};
        return err;
    };

    if (std.fs.path.dirname(dst_path)) |dir_path| {
        helpers.fsyncDir(dir_path);
    }
}

pub fn ensureParentDir(path: []const u8) !void {
    const dir_path = std.fs.path.dirname(path) orelse return;
    std.fs.makeDirAbsolute(dir_path) catch |err| switch (err) {
        error.PathAlreadyExists => return,
        error.FileNotFound => {
            try ensureParentDir(dir_path);
            std.fs.makeDirAbsolute(dir_path) catch |e| switch (e) {
                error.PathAlreadyExists => return,
                else => return e,
            };
        },
        else => return err,
    };
}

pub fn removeEmptyParentDirs(path: []const u8, stop_at: []const u8) void {
    var current = std.fs.path.dirname(path);
    while (current) |dir| {
        if (dir.len <= stop_at.len) break;
        std.fs.deleteDirAbsolute(dir) catch break;
        current = std.fs.path.dirname(dir);
    }
}

File: src/replication.rs
//! Replication worker: processes the replication log queue.
//!
//! Each worker thread calls `repl_log.wait_next()` in a loop and either
//! copies the file to the replica (`put`) or removes it (`delete`).

use std::fs;
use std::io::{self, Read, Write};
use std::os::unix::fs::{MetadataExt, PermissionsExt};
use std::path::Path;
use std::sync::atomic::Ordering;
use std::sync::Arc;

use crate::helpers;
use crate::repl_log::ReplOp;
use crate::state::FsState;

/// Run the replication worker loop. Returns when `state.shutting_down` is set
/// and there are no more pending entries.
pub fn replication_worker(state: &Arc<FsState>) {
    log::info!("Replication worker started");
    loop {
        // Try to get next entry. wait_next() returns None after a timeout
        // if no entries are pending, allowing us to check shutdown.
        let (id, entry) = match state.repl_log.wait_next() {
            Some(pair) => pair,
            None => {
                if state.shutting_down.load(Ordering::Relaxed) {
                    // Drain any stragglers
                    while let Some((id, entry)) = state.repl_log.try_next() {
                        let _ = match entry.op {
                            ReplOp::Put => replicate_put(state, &entry.path),
                            ReplOp::Delete => replicate_delete(state, &entry.path),
                        };
                        state.repl_log.mark_completed(id);
                    }
                    break;
                }
                continue;
            }
        };

        let result = match entry.op {
            ReplOp::Put => replicate_put(state, &entry.path),
            ReplOp::Delete => replicate_delete(state, &entry.path),
        };

        match result {
            Ok(()) => {
                log::debug!("Replicated {:?} {}", entry.op, entry.path);
            }
            Err(e) => {
                log::error!("Replication failed for {:?} {}: {}", entry.op, entry.path, e);
            }
        }

        state.repl_log.mark_completed(id);
    }
    log::info!("Replication worker stopped");
}

/// Copy a file (and its .sum sidecar) from backing to replica.
fn replicate_put(state: &FsState, rel: &str) -> io::Result<()> {
    let src = state.backing_path(rel);
    let dst = state.replica_file_path(rel);

    // Handle symlinks
    let meta = fs::symlink_metadata(&src)?;
    if meta.file_type().is_symlink() {
        let target = fs::read_link(&src)?;
        // Ensure parent dir exists in replica
        if let Some(parent) = dst.parent() {
            fs::create_dir_all(parent)?;
        }
        // Remove any existing file/symlink at destination
        let _ = fs::remove_file(&dst);
        std::os::unix::fs::symlink(&target, &dst)?;
        return Ok(());
    }

    if !meta.is_file() {
        return Ok(()); // Skip directories, etc.
    }

    // Ensure parent dir exists in replica
    if let Some(parent) = dst.parent() {
        fs::create_dir_all(parent)?;
    }

    // Copy file content with sync
    copy_file_with_sync(&src, &dst)?;

    // Preserve permissions
    let perms = fs::Permissions::from_mode(meta.mode() & 0o7777);
    fs::set_permissions(&dst, perms)?;

    // Copy .sum sidecar if it exists
    let src_sum = helpers::sum_path_for(&src);
    let dst_sum = helpers::sum_path_for(&dst);
    if src_sum.exists() {
        copy_file_with_sync(&src_sum, &dst_sum)?;
    }

    Ok(())
}

/// Remove a file (and its .sum sidecar) from the replica.
///
/// Guard: if the backing file currently exists with a `.sum` sidecar, the file
/// was re-created (and checksummed) since this delete entry was enqueued.
/// Deleting the replica copy would lose the latest version, so we skip the
/// delete.  With multiple worker threads, this prevents an older delete from
/// executing after a newer put for the same path has already replicated the
/// re-created file.  (In normal operation, FUSE `unlink` removes the backing
/// file and .sum before enqueuing the delete, so this guard only fires when
/// the file was truly re-created.)
fn replicate_delete(state: &FsState, rel: &str) -> io::Result<()> {
    let src = state.backing_path(rel);
    let src_sum = helpers::sum_path_for(&src);
    if src.exists() && src_sum.exists() {
        log::debug!(
            "Skipping replica delete for {} — file was re-created",
            rel
        );
        return Ok(());
    }

    let dst = state.replica_file_path(rel);
    let dst_sum = helpers::sum_path_for(&dst);

    // Remove file — ignore NotFound
    match fs::remove_file(&dst) {
        Ok(()) => {}
        Err(e) if e.kind() == io::ErrorKind::NotFound => {}
        Err(e) => return Err(e),
    }

    // Remove .sum sidecar — ignore NotFound
    match fs::remove_file(&dst_sum) {
        Ok(()) => {}
        Err(e) if e.kind() == io::ErrorKind::NotFound => {}
        Err(e) => return Err(e),
    }

    // Try to remove empty parent dirs (best effort, like Zig version)
    if let Some(parent) = dst.parent() {
        remove_empty_parents(parent, &state.replica_dir.join("files"));
    }

    Ok(())
}

/// Copy `src` to `dst` via temp file + rename, with fsync.
pub fn copy_file_with_sync(src: &Path, dst: &Path) -> io::Result<()> {
    let mut tmp_name = dst.as_os_str().to_os_string();
    tmp_name.push(".helmetfs-tmp");
    let tmp = std::path::PathBuf::from(tmp_name);
    {
        let mut reader = fs::File::open(src)?;
        let mut writer = fs::File::create(&tmp)?;
        let mut buf = [0u8; 65536];
        loop {
            let n = reader.read(&mut buf)?;
            if n == 0 {
                break;
            }
            writer.write_all(&buf[..n])?;
        }
        writer.sync_all()?;
    }
    fs::rename(&tmp, dst)?;
    Ok(())
}

/// Remove empty parent directories up to (but not including) `stop_at`.
fn remove_empty_parents(mut dir: &Path, stop_at: &Path) {
    while dir != stop_at {
        if fs::remove_dir(dir).is_err() {
            break;
        }
        match dir.parent() {
            Some(p) => dir = p,
            None => break,
        }
    }
}
